src/Host/Client/Impl/Host/RemoteRHostConnector.cs

```diff
@@ -3,32 +3,68 @@
     internal sealed class RemoteRHostConnector : RHostConnector {
+        private readonly NetworkCredential _credentials;
+        private readonly AutoResetEvent _credentialsValidated = new AutoResetEvent(true);
+        private readonly string _authority;
+        private bool _ignoreSavedCredentials;
+
         public RemoteRHostConnector(Uri brokerUri)
             : base(brokerUri.Fragment) {
 
-            CreateHttpClient();
-            Broker.BaseAddress = brokerUri;
+            _credentials = new NetworkCredential();
+            _authority = new UriBuilder() { Scheme = brokerUri.Scheme, Host = brokerUri.Host, Port = brokerUri.Port }.ToString();
+
+            CreateHttpClient(brokerUri, _credentials);
         }
 
-        protected override HttpClientHandler GetHttpClientHandler() {
-            return new HttpClientHandler {
-                Credentials = new UICredentials()
-            };
+        protected override Task ConnectToBrokerAsync() {
+            return Task.CompletedTask;
+        }
+
+        private void GetCredentials(out string userName, out string password) {
+            // If there is already a GetCredentials request for which there hasn't been a validation yet, wait until it completes.
+            // This can happen when two sessions are being created concurrently, and we don't want to pop the credential prompt twice -
+            // the first prompt should be validated and saved, and then the same credentials will be reused for the second session.
+            _credentialsValidated.WaitOne();
+
+            var userNameBuilder = new StringBuilder(CREDUI_MAX_USERNAME_LENGTH + 1);
+            var passwordBuilder = new StringBuilder(CREDUI_MAX_PASSWORD_LENGTH + 1);
+
+            bool save = false;
+
+            int flags = CREDUI_FLAGS_EXCLUDE_CERTIFICATES | CREDUI_FLAGS_PERSIST | CREDUI_FLAGS_EXPECT_CONFIRMATION | CREDUI_FLAGS_GENERIC_CREDENTIALS;
+            if (_ignoreSavedCredentials) {
+                flags |= CREDUI_FLAGS_ALWAYS_SHOW_UI;
+            }
+
+            int err = CredUIPromptForCredentials(IntPtr.Zero, _authority, IntPtr.Zero, 0, userNameBuilder, userNameBuilder.Capacity, passwordBuilder, passwordBuilder.Capacity, ref save, flags);
+            if (err != 0) {
+                throw new OperationCanceledException("No credentials entered.");
+            }
+
+            userName = userNameBuilder.ToString();
+            password = passwordBuilder.ToString();
         }
 
-        protected override void ConfigureWebSocketRequest(HttpWebRequest request) {
-            request.AuthenticationLevel = AuthenticationLevel.MutualAuthRequested;
-            request.Credentials = CredentialCache.DefaultNetworkCredentials;
+        protected override void UpdateCredentials() {
+            string userName, password;
+            GetCredentials(out userName, out password);
+
+            _credentials.UserName = userName;
+            _credentials.Password = password;
```

Comment:
It's intentional - if they were incorrect, this will change them to the new user input after asking the user.

Changing them directly on the instance of `NetworkCredential` is kinda weird, but it seems to be the only way - `HttpClientHandler.Credentials` is read-only once the first request is made, so the only way to change credentials afterwards is by mutating the instance directly here.